cou/steps/ceph.py

```diff
@@ -0,0 +1,279 @@
+    """
+    ceph_mon_units = list(ceph_mon_app.units.values())
+    if not ceph_mon_units:
+        raise UnitNotFound(f"'{ceph_mon_app.name}' has no units, is this a valid OpenStack cloud?")
+    return ceph_mon_units[0].name
+
+
+async def osd_noout(model: Model, enable: bool) -> None:
+    """Set or unset 'noout' for ceph cluster(s).
+
+    Note this will set or unset 'noout' flag for all ceph clusters present in
+    the model.
+
+    :param model: The juju model to work with
+    :type model: Model
+    :param enable: True to set noout, False to unset noout
+    :type enable: bool
+    """
+    try:
+        ceph_mon_apps = await _get_applications(model)
+    except ApplicationNotFound as e:
+        logger.warning("%s", str(e))
+        logger.warning("Skip changing 'noout', because there's no ceph-mon applications.")
+        return
+
+    for ceph_mon_app in ceph_mon_apps:
+        try:
+            ceph_mon_unit_name = await _get_unit_name(ceph_mon_app)
+        except UnitNotFound as e:
+            logger.warning("%s", str(e))
+            logger.warning(
+                "Skip changing 'noout', because there's no %s units.", ceph_mon_app.name
+            )
+            continue
+
+        await model.run_action(
+            ceph_mon_unit_name, f"{'set' if enable else 'unset'}-noout", raise_on_failure=True
+        )
+
+
+async def get_osd_noout_state(model: Model, unit_name: str) -> bool:
+    """Check if noout is set on ceph cluster or not.
+
+    :param model: The juju model to work with
+    :type model: Model
+    :param unit_name: The name of the unit
+    :type unit_name: str
+    :return: True if noout is set, otherwise False
+    :type: bool
+    """
+    results = await model.run_on_unit(unit_name, "ceph osd dump -f json")
+    return "noout" in json.loads(results["stdout"].strip()).get("flags_set", [])
+
+
+async def assert_osd_noout_state(model: Model, state: bool) -> None:
+    """Assert ceph cluster is set (state=True) or unset (state=False).
+
+    Note this will assert 'noout' flag is in the desired state for all ceph
+    clusters present in the model.
+
+    :param model: The juju model to work with
+    :type model: Model
+    :param state: True noout is set, otherwise False
+    :type state: bool
+    :raise ApplicationError: if noout is not in desired state
+    """
+    try:
+        ceph_mon_apps = await _get_applications(model)
+    except ApplicationNotFound as e:
+        logger.warning("%s", str(e))
+        logger.warning("Skip verifying 'noout', because there's no ceph-mon applications.")
+        return
+
+    error = False
+    for ceph_mon_app in ceph_mon_apps:
+        try:
+            ceph_mon_unit_name = await _get_unit_name(ceph_mon_app)
+        except UnitNotFound as e:
+            logger.warning("%s", str(e))
+            logger.warning(
+                "Skip verifying 'noout', because there's no %s units.", ceph_mon_app.name
+            )
+            continue
```

Comment:
for `[app for app in apps if isinstance(app, CephMon)]`  vs `[app for app in apps.values() if app.charm == "ceph-mon"]` 
and the helper `_get_applications` and `_get_unit_name`, yes. That can be in `app_utils.py` though, if it's used in other place. So making it as a helper function does not hurt. We can refactor that as we see needed

But `_get_ceph_mon_post_upgrade_steps` is a refactoring: I simply inline this step in `_get_post_upgrade_steps`

Reply:
For example, `vault.py` and `nova_cloud_controller.py` uses a similar approach to get unit / app.

Reply:
> I don't agree with the vault and nova_cloud_controller approach and I think they can be refactored later.
Filtering by the COU applications would be better because it would use less the model once we already have this information already processed from the analysis_result.

This is the point. I think it's not important to do a correct way right here, because the effort will be removed after the refactoring. Most likely this will be an function in `app_utils.py` e.g. `get_application_by_class` or some thing, and the line you commented will be gone.

Reply:
Are you okay with I create a follow up PR for refactoring this part only?

Reply:
Yes I'm ok with that. My concern is that we keep increasing tech debt

Reply:
That's true, we are aware of that months ago. And we finally plan to do some clean up / small refactoring in the coming week